### ISSUE TITLE
Freezing + DB write errors with WebDav

### ISSUE THREAD

**Issue by the-moog** [keepassxreboot/keepassxc#9491]
## Overview

Evaluating KeePass and having immediate issues with updating records.
Linux client (testing only, not live)
Only one client, this machine
Client ...er... client = Firefox ESR
KeePass install from Debian package

DB File is remote on WebDav
(Being The point of the eval - for platform agnostic client capabilities)
Currently Dav is mounted using Fuse as I could not get KeePass to accent a URL

### First thoughts?
Looks to me like a classic race condition - against itself?

## Steps to Reproduce
[NOTE]: # ( Provide a simple set of steps to reproduce this bug. )
1. Do something that modifies a/some record(s)
2. Don't wait very long < 5~15s perhaps
3. Do something else that modifies a record

Could be that doing something that modifies more than one record makes this appear more.
Was having lots occurrences after drag-dropping groups of records (5-10) at a time into sub-groups.

## Expected Behavior
No warnings
No hangs

## Actual Behavior
Get warnings on record write operations
Usually followed by application freeze

### Out of data version?
Am I just using an 'old' version?
Using : 2.6.2
Latest : 2.7.2 - Does this have improvements related to DB writes?

### Questions:
I can see from `top` that multiple threads in the client. Is it possible more than one could both try and write a the same time?
How does the file and record locking work?
What if the backend does any sort of file cacheing?
e.g. write1 - read1 - write2 - read2 may read the backend data from write1 when read2 occurs if the write2 is not committed yet.
Is the database transactional?

Is there any way to use WebDav directly rather than via a mount?  ( I tried and could not get it to accept a URL)
Could it be that the mount hides the inherent resource locking in WebDav and the versioning that comes from DeltaV?

### Related
After the message, selecting either option "don't save changes" or "merge changes" client hangs, the GUI freezes - glaring perhaps?
When this freeze happens and KeePass is run from the command line I see hundreds of lines like this.
`Maximum depth of replacement has been reached. Entry uuid: {<LOTS OF HEX DIGITS IN UUID FORM>}`
After the hang the connection to Firefox is lost (hints of a blocking thread - timeouts/retries not implemented?)

Quick scan of possibly similar or identical issues #9371 #9367 #9340 #9181

#### Logging bug or missing feature
I note that when I get issues I see errors like the `Maximum depth...` and `Entry of <name>
[<string of hex>] contains conflicting changes - conflict resolution may lose data!`
And this relates to but is not the same text as errors reported via the GUI.
The on screen GUI errors are not always output on the console.
It looks like logging is not unified - omission?

[NOTE]: # ( Paste debug info from Help → About here )
$ keepassxc --debug-info rather than screen grab

KeePassXC - Version 2.6.2
Revision: e9b9582

Qt 5.15.2
Debugging mode is disabled.

Operating system: Debian GNU/Linux 11 (bullseye)
CPU architecture: x86_64
Kernel: linux 5.15.103-1-x86_64

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (only unsigned sharing)
- YubiKey
- Secret Service Integration

Cryptographic libraries:
 libgcrypt 1.8.8

[NOTE]: # ( Pick choices based on your environment )
Operating System: Linux-Debian
Desktop Env: KDE
Windowing System: X11

**Comment by droidmonkey** [comment]
First, update to 2.7.5 (use Flatpak if you have to). However, we don't necessarily control the saving operation outside of telling qt to write a file. FUSE is notoriously terrible. After you update, you should investigate using one of the following options: https://keepassxc.org/docs/KeePassXC_UserGuide#_advanced_save_options

**Comment by the-moog** [comment]
Thanks for the pointer to the options, will look at that.  I really did not want to use FUSE and was investigating how well it works with WebDav - Fuse was a compromise to allow the test to continue.  I was put native WebDav support down to me not understanding the dialogs (I've seen a few odd dialogs) e.g. Create database gets confused and asks you to enter a filename but grays out the save button - In fact, it's just using the wrong native dialog - it should be select folder, so it's the mapping to the Qt style that's wrong.

**Comment by the-moog** [comment]
I tried those settings on 2.6 and it did not work.   I did a large update (group drag of ~20) waited well over a minute then did a single drag and it hung.  Will Try on 2.7.  Is 2.6 deprecated?

**Comment by droidmonkey** [comment]
We don't support any versions less than the current one. Dragging one or 20 entries should be irrelevant to saving. We have a delay prior to kicking off autosave that allows for multiple changes to be completed. All faults here likely lie with the FUSE interface.

**Comment by the-moog** [comment]
@droidmonkey So if fuse is 'broke' (this is the only time I've had so much trouble as been using it with Nextcloud and subversion for years) as mentioned initially, can KeePass use Dav directly?
Only in my system it seems to be using the theme dialogs and then it's using the one for opening files that is lacking a way to enter a URL.  So it can only open local files, hence trying a mount....
Ok: just found #1775 so I guess the answer is no, unless somebody is working on this in a fork?

**Comment by the-moog** [comment]
Wow that locked up good that time.  Took out an entire VM.  Not seen that sort of lockup in years.
Got a zombie keepass somehow re-attached to the systemd process (which I think is not true, but something in the process tree got very corrupted) so I could not kill it off...  Look at /proc and it was not there.  Had to hard kill the VM to shut it down. 

So, It seems the suggestion of backups made it worse.  I think there are glaring threads so two IO threads are waiting on each other blocking all other IO threads. Add backups and you get more queued IO and more chance of that happening. Would needs a debugger to tell for sure. But at least it's easy to reproduce.

BTW: My experience is that Dav only get a bad name because people don't set it up right and so many implementations are (or were) half baked.  The early days of SVN were misery.  Later I was using it with 10's of thousands files, some several GB each no problems, ever.

### Prediction
Looking into this, I can see ways you can break KeePass with the current defaults on davfs2.

It the filesystem (or the route through it) tries to minimize the writes using a lot of caching and if KeePass works by re-opening files between writes then I can see ways it could screw up.  As it may assume the last write is on the disk and it might not be yet.

If KeePass makes assumptions about the filesystem state, I would guess this does not matter what the underlying filesystem is.  That may explain why there similar reported issues; I.e. where there is caching going on and the filesystem driver takes time to work. The likes of OneDrive, GoogleDrive etc.

### Better Resolution
A good solution may be to use something like a counting semaphore, stored in the database. But turning off caching should help for now, lets try it....

### Temporary Resolution
So:
- Turn off caching
- ensure locks are enabled (it seems they are by default)
- Quicker timeouts and retries so we can check the lock more often

I put the following settings in /etc/davfs2/davfs2.conf
```
# Compress data: CPUs are fast and have H/W AES.
use_compression 1
# Retry quicker (so more retries)
retry 5
# Only wait a short time to re-scan the dir
dir_refresh     10
# Never wait and return the cached file always immediately get remote
file_refresh 0
# Never wait to commit cache
delay_upload 0
```
I've not had a lockup since.....  Yay.  Oh I updated to 2.7 too, but I can make it die in that if I revert the settings for dsvfs2.

**Comment by droidmonkey** [comment]
Good thing you found the error, poorly behaving webdav software is unfortunately common. The atomic saves likely do not jive at all with the "smart" behavior of these network file handlers. It seems like it's a keepassxc problem, but in reality any software using QSaveFile from Qt would experience the same issues. Not much software actually does that though because they all just write to the same file.

**Comment by the-moog** [comment]
@droidmonkey Just to let you know I've now got KeePass in Linux to work well with Android using Keepass2Android over webdav/s and it's working really well. Windows and WSL next....
It seems Android on my S23U (via a cloud proxy to a QNAP NAS on the LAN) over 4G is about 10x faster than Linux when it's on  Gigabit LAN with a 10Gb connection NAS to Router. Very odd....

Hopefully you got some useful data here?  It looks like a bunch of outstanding issues are likely very closely related. Maybe these results will trigger some ideas.  e.g. Could be related to a Qt5 bug reported in 5.6 and [not 'fixed' until 5.11](https://bugreports.qt.io/browse/QTBUG-57299)
